src/Compilers/Core/Portable/Syntax/SyntaxNode.cs

```diff
@@ -426,24 +432,81 @@ public bool ContainsSkippedText
         }
 
         /// <summary>
-        /// Determines whether this node has any descendant preprocessor directives.
+        /// Determines whether this node or any of its descendant nodes, tokens or trivia have any diagnostics on them. 
         /// </summary>
-        public bool ContainsDirectives
+        public bool ContainsDiagnostics
         {
             get
             {
-                return this.Green.ContainsDirectives;
+                return this.Green.ContainsDiagnostics;
             }
         }
 
         /// <summary>
-        /// Determines whether this node or any of its descendant nodes, tokens or trivia have any diagnostics on them. 
+        /// Determines whether this node has any descendant preprocessor directives.
         /// </summary>
-        public bool ContainsDiagnostics
+        public bool ContainsDirectives => this.Green.ContainsDirectives;
+
+        /// <summary>
+        /// Returns true if this node contains any conditional directives (<c>#if</c> in C# and <c>#If</c> in Visual
+        /// Basic) within it.  This is similar to <see cref="ContainsDiagnostics"/> except that where that returns true
+        /// if there are any preprocessor directives whatsoever, this is only true for for <c>if</c> directives.
+        /// </summary>
+        public bool ContainsConditionalDirectives()
```

Comment:
- Should the result be cached in case this is called multiple times?
- Is there a way to avoid the traversal and mark the node (and parent chain) as `ContainsConditionalDirectives` while parsing?
- Is it beneficial to know the **exact** conditional directives used, and have the possibility of sharing the tree based on `parseOptions.PreprocessorSymbolNames`, something like the following logic (Note: the logic isn't completely correct as it doesn't take `#define` into account - just showing the idea):

    ```csharp
    bool CanShareTree(ParseOptions oldTreeOptions, ParseOptions newTreeOptions, ImmutableArray<string> directivesInTree)
    {
        var oldPreprocessors = oldTreeOptions.PreprocessorSymbolNames;
        var newPreprocessors = newTreeOptions.PreprocessorSymbolNames;
        foreach (var directiveInTree in directiveInTrees)
        {
            if ((oldPreprocessors.Contains(directiveInTree) ^ newPreprocessors.Contains(directiveInTree)) == 1)
                return false;
        }

        return true;
    }
    ```

Can we get the number of extra trees that will be shared with this idea for Roslyn.sln? If it's a good number, maybe worth investigating the idea?

Reply:
> Should the result be cached in case this is called multiple times?

Caching would require additional space and I'm not at all comfortable increasing the size of green nodes. 

> Is there a way to avoid the traversal and mark the node (and parent chain) as ContainsConditionalDirectives while parsing?

Same issue as above. Also included in he API review alternatives though.

Reply:
> Can we get the number of extra trees that will be shared with this idea for Roslyn.sln? If it's a good number, maybe worth investigating the idea?

Sure. I can collect that.